Mocking SQL Statement causes test to fail, when using AppMap Maven plugin

I am using the AppMap Maven plugin to instrument my builds. I have a JUnit test that forces an error in the execution of a SQL query. The test is set up with Mockito like this:

```java
Statement statement = mock(Statement.class);
when(statement.execute(anyString()))
  .thenThrow(new SQLException("Exception executing SQL statement"));
```

My build fails when AppMap tries to introspect the database name. The code at `com.appland.appmap.process.hooks.SqlQuery.getDbName(SqlQuery.java:47)` expects a `SQLException`, but an `InvalidUseOfMatchersException` exception is thrown instead, failing the build. More details are in the stack trace below. 

```
[ERROR] us.fatehi.utility.test.DatabaseUtilityTest.executeSql_throw  Time elapsed: 0.433 s  <<< ERROR!
org.mockito.exceptions.misusing.InvalidUseOfMatchersException:

Invalid use of argument matchers!
0 matchers expected, 1 recorded:
-> at us.fatehi.utility.test.DatabaseUtilityTest.executeSql_throw(DatabaseUtilityTest.java:147)

This exception may occur if matchers are combined with raw values:
    //incorrect:
    someMethod(any(), "raw String");
When using matchers, all arguments have to be provided by matchers.
For example:
    //correct:
    someMethod(any(), eq("String by matcher"));

For more info see javadoc for Matchers class.

        at com.appland.appmap.process.hooks.SqlQuery.getDbName(SqlQuery.java:47)
        at com.appland.appmap.process.hooks.SqlQuery.recordSql(SqlQuery.java:61)
        at com.appland.appmap.process.hooks.SqlQuery.execute(SqlQuery.java:122)
        at us.fatehi.utility.test.DatabaseUtilityTest.executeSql_throw(DatabaseUtilityTest.java:147)
		...
```

--- a/agent/src/main/java/com/appland/appmap/process/hooks/SqlQuery.java
+++ b/agent/src/main/java/com/appland/appmap/process/hooks/SqlQuery.java
@@ -9,6 +9,7 @@ import com.appland.appmap.transform.annotations.Unique;
 import com.appland.appmap.util.Logger;
 
 import java.sql.Connection;
+import java.sql.DatabaseMetaData;
 import java.sql.SQLException;
 import java.sql.Statement;
 
@@ -29,10 +30,23 @@ public class SqlQuery {
     recorder.add(event);
   }
 
+  private static boolean isMock(Object o) {
+    return o.getClass().getPackage().getName().startsWith("org.mockito");
+  }
+
   private static String getDbName(Connection c) {
     String dbname = "";
+    if (c == null) {
+      return dbname;
+    }
+
     try {
-      dbname = c.getMetaData().getDatabaseProductName();
+      DatabaseMetaData metadata;
+      if (isMock(c) || isMock(metadata = c.getMetaData())) {
+        return "[mocked]";
+      }
+
+      dbname = metadata.getDatabaseProductName();
     }
     catch (SQLException e) {
       Logger.println("WARNING, failed to get database name");
@@ -43,7 +57,15 @@ public class SqlQuery {
 
   private static String getDbName(Statement s) {
     String dbname = "";
+    if (s == null) {
+      return dbname;
+    }
+
     try {
+      if(isMock(s)) {
+        return "[mocked]";
+      }
+
       dbname = getDbName(s.getConnection());
     }
     catch (SQLException e) {

--- a/agent/src/test/java/com/appland/appmap/integration/MockerTest.java
+++ b/agent/src/test/java/com/appland/appmap/integration/MockerTest.java
@@ -1,0 +1,39 @@
+package com.appland.appmap.integration;
+
+import org.junit.Test;
+
+import static org.mockito.Mockito.mock;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+// These tests are run with recording enabled. Getting to the end of the test
+// shows that recording is working correctly.
+public class MockerTest {
+  @Test
+  public void testMockedStatement() {
+    final Statement s = mock(Statement.class);
+    try {
+      s.execute("select 1;");
+    }
+    catch (SQLException e) {
+      e.printStackTrace();
+    }
+
+    assert(true);
+  }
+
+  @Test
+  public void testMockedConnection() {
+    final Connection c = mock(Connection.class);
+    try {
+      final String s = c.nativeSQL("select 1");
+    }
+    catch (SQLException e) {
+      e.printStackTrace();
+    }
+
+    assert(true);
+  }
+}

